Azure: check for SAS in all targets

JAVGan opened the pull request:
This commit refactors the `AzureService.publish` to look up for the existing SAS uri in all targets (`preview`, `live`, `draft`) before assuming it's not present and making changes.

It keeps the behavior of `overwrite` which always adds the new SAS URI to the plan, regardless of finding it or not.

Refers to SPSTRAT-595

lslebodn commented:
@JAVGan I tried to review the PR but I am not sure whether I understand enough the order returned by `compute_targets()`
```
>>> compute_targets()
['preview', 'live', 'draft']
>>> compute_targets('preview')
['preview', 'live', 'draft']
>>> compute_targets('live')
['live', 'preview', 'draft']
>>> compute_targets('draft')
['draft', 'preview', 'live']
```

which is quite essential to understand the changes in the MR.

My expectation for checking SAS uri would be:
1) SAS uri is in live => do nothing
2) SAS uri is in preview => publish to live
3) SAS uri is in draft => publish to preview && publish to live
4) missing in all => attach to draft && publish to preview && publish to live

It is quite oversimplified and maybe even state diagram would help.
We might also discuss changes in some online meeting rather than in PR
We  would provide summary here. (just in case it would be faster :-)

JAVGan commented:
@lslebodn feel free to schedule a meeting with me whenever you have time then we can go through it together

JAVGan commented:
@lslebodn I've removed the nested loops and now all methods `get_product_*` requires the proper submission `target` to retrieve data.

I also updated the `publish` to use the targets from `compute_targets` and then the loop is simplified, without all that unnecessary calls.

The new commits are:

- 0fe3ad89399a68adb589967535bcd2c0a8fdca48 
- 5231b1731e2c8adf6a6f64ecf149d1378195fc72

Please let me know WDYT

ashwgit commented:
It looks good to me.

lslebodn commented on cloudpub/ms_azure/service.py:
```diff
@@ -556,6 +566,31 @@ def get_modular_resources_to_publish(
+        Returns:
+            List[Str]: The ordered list with targets to lookup.
+        """
+        all_targets = ["preview", "live", "draft"]
```
Why is `preview` before `live`?

JAVGan replied:
Because we need to first iterate over preview to detect faster if the SAS is already in preview. It was part of already existing logic which I opted to keep it.

lslebodn commented on cloudpub/ms_azure/service.py:
```diff
@@ -241,7 +243,21 @@ def list_products(self) -> List[ProductSummary]:
+            The requested product ID when found.
+        Raises NotFoundError when the product was not found.
+        """
+        for product in self.list_products():
```
There is caching in the slef.list_products
* https://github.com/release-engineering/cloudpub/blob/0fa3a5ca3c6b2ad729ce05d0aae9709a33bc30a9/cloudpub/ms_azure/service.py#L242-243

I am not sure whether it is good idea to cache such data.
Is there any chance it could cause issues when we try to gracefully handle updates to the same offer from different processes (pub tasks)?

JAVGan commented on cloudpub/ms_azure/service.py:
```diff
@@ -664,75 +816,57 @@ def publish(self, metadata: AzurePublishingMetadata) -> None:
+                tech_config = res["tech_config"]
+                # We don't want to seek for SAS anymore as it was already found
+                if res["sas_found"]:
+                    break
```
@lslebodn Fix is here: https://github.com/release-engineering/cloudpub/pull/146

@ashwgit PTAL as well
